File: apps/api/src/common/featureManifest.ts
import type { driverPredictionFeatures, driverSprintFeatures } from '../db/schema';
import type { FeatureScores, SprintFeatureScores } from './types';

// apps/api never computes with feature weights — it only reads and re-labels stored scores
// (weighting happens in data-engine's compute_features.py / compute_sprint_features.py, and
// in apps/web's predictionMath.ts for the interactive What-If lab). So this manifest carries
// no weight field; it just dedupes the key/column/label/nullable mapping that used to be
// hand-typed separately in toFeatures(), toSprintFeatures(), and intel-standings.helpers.ts's
// FEATURE_COLUMNS. `column: keyof TRow` makes a manifest entry referencing a column that
// doesn't exist on the Drizzle-inferred row type a compile error, not silent drift; `TKey`
// pins `key` to the target FeatureScores/SprintFeatureScores type so consumers that index a
// Record<keyof FeatureScores, ...> by manifest key stay fully type-checked too.
export interface FeatureManifestEntry<TRow, TKey extends string = string> {
  key: TKey;
  column: keyof TRow;
  label: string;
  nullable: boolean;
}

export const GP_FEATURE_MANIFEST: FeatureManifestEntry<typeof driverPredictionFeatures.$inferSelect, keyof FeatureScores>[] = [
  { key: 'carPerformance', column: 'carPerformanceScore', label: 'Car Performance', nullable: false },
  { key: 'driverRating', column: 'driverRatingScore', label: 'Driver Rating', nullable: false },
  { key: 'startingPosition', column: 'startingPositionScore', label: 'Starting Position', nullable: false },
  { key: 'winRate', column: 'winRateScore', label: 'Win Rate', nullable: false },
  { key: 'luckFactor', column: 'luckFactorScore', label: 'Luck Factor', nullable: false },
  { key: 'weatherImpact', column: 'weatherImpactScore', label: 'Weather Impact', nullable: false },
  { key: 'trackOvertake', column: 'trackOvertakeScore', label: 'Track Overtake', nullable: true },
  { key: 'positionGain', column: 'positionGainScore', label: 'Position Gain', nullable: false },
  { key: 'longRunPace', column: 'longRunPaceScore', label: 'Long Run Pace', nullable: true },
  { key: 'reliability', column: 'reliabilityScore', label: 'Reliability', nullable: true },
  { key: 'qualifyingDelta', column: 'qualifyingDeltaScore', label: 'Qualifying Delta', nullable: true },
  { key: 'sectorStrength', column: 'sectorStrengthScore', label: 'Sector Strength', nullable: true },
  { key: 'tyreDeg', column: 'tyreDegScore', label: 'Tyre Degradation', nullable: true },
  { key: 'circuitAdjStartPos', column: 'circuitAdjStartPosScore', label: 'Circuit-Adj. Start Position', nullable: true },
  { key: 'circuitAdjPositionGain', column: 'circuitAdjPositionGainScore', label: 'Circuit-Adj. Position Gain', nullable: true },
];

export const SPRINT_FEATURE_MANIFEST: FeatureManifestEntry<typeof driverSprintFeatures.$inferSelect, keyof SprintFeatureScores>[] = [
  { key: 'carPerformance', column: 'carPerformanceScore', label: 'Car Performance', nullable: false },
  { key: 'startingPosition', column: 'startingPositionScore', label: 'Starting Position', nullable: false },
  { key: 'driverRating', column: 'driverRatingScore', label: 'Driver Rating', nullable: false },
  { key: 'trackOvertake', column: 'trackOvertakeScore', label: 'Track Overtake', nullable: true },
  { key: 'shortRunPace', column: 'shortRunPaceScore', label: 'Short Run Pace', nullable: false },
  { key: 'circuitAdjStartPos', column: 'circuitAdjStartPosScore', label: 'Circuit-Adj. Start Position', nullable: true },
  { key: 'sqQualifyingDelta', column: 'sqQualifyingDeltaScore', label: 'SQ vs Teammate', nullable: true },
  { key: 'weatherImpact', column: 'weatherImpactScore', label: 'Weather Impact', nullable: false },
  { key: 'winRate', column: 'winRateScore', label: 'Win Rate', nullable: false },
  { key: 'luckFactor', column: 'luckFactorScore', label: 'Luck Factor', nullable: false },
];

export function mapFeatureRow<TRow extends Record<string, unknown>, TOut>(
  raw: TRow,
  manifest: FeatureManifestEntry<TRow>[],
): TOut {
  const out: Record<string, unknown> = {};
  for (const { key, column, nullable } of manifest) {
    out[key] = nullable ? (raw[column] ?? null) : raw[column];
  }
  return out as TOut;
}

File: data-engine/src/utils/feature_manifest.py
from dataclasses import dataclass


@dataclass(frozen=True)
class FeatureSpec:
    name: str
    weight: float
    label: str
    nullable: bool


GP_FEATURES: tuple[FeatureSpec, ...] = (
    FeatureSpec("car_performance", 0.20, "Car Performance", nullable=False),
    FeatureSpec("long_run_pace", 0.15, "Long Run Pace", nullable=True),
    FeatureSpec("tyre_deg", 0.08, "Tyre Degradation", nullable=True),
    FeatureSpec("reliability", 0.08, "Reliability", nullable=True),
    FeatureSpec("qualifying_delta", 0.08, "Qualifying Delta", nullable=True),
    FeatureSpec("driver_rating", 0.08, "Driver Rating", nullable=False),
    FeatureSpec("win_rate", 0.08, "Win Rate", nullable=False),
    FeatureSpec("luck_factor", 0.07, "Luck Factor", nullable=False),
    FeatureSpec("sector_strength", 0.06, "Sector Strength", nullable=True),
    FeatureSpec("circuit_adj_start_pos", 0.07, "Circuit-Adj. Starting Position", nullable=True),
    FeatureSpec("circuit_adj_position_gain", 0.03, "Circuit-Adj. Position Gain", nullable=True),
    FeatureSpec("weather_impact", 0.02, "Weather Impact", nullable=False),
)
GP_WEIGHTS: dict[str, float] = {f.name: f.weight for f in GP_FEATURES}

SPRINT_FEATURES: tuple[FeatureSpec, ...] = (
    FeatureSpec("car_performance", 0.25, "Car Performance", nullable=False),
    FeatureSpec("circuit_adj_start_pos", 0.25, "Circuit-Adj. Starting Position", nullable=True),
    FeatureSpec("short_run_pace", 0.10, "Short Run Pace", nullable=False),
    FeatureSpec("driver_rating", 0.10, "Driver Rating", nullable=False),
    FeatureSpec("weather_impact", 0.08, "Weather Impact", nullable=False),
    FeatureSpec("win_rate", 0.08, "Win Rate", nullable=False),
    FeatureSpec("luck_factor", 0.08, "Luck Factor", nullable=False),
    FeatureSpec("sq_qualifying_delta", 0.06, "SQ vs Teammate", nullable=True),
)
SPRINT_WEIGHTS: dict[str, float] = {f.name: f.weight for f in SPRINT_FEATURES}

SOFTMAX_TEMPERATURE = 0.3


def assemble_scores(values: dict[str, float], features: tuple[FeatureSpec, ...]) -> dict[str, float]:
    """Raises KeyError if a manifest feature's score was never computed, instead of
    silently summing an incomplete weighted score."""
    return {f.name: values[f.name] for f in features}
